adding a very simple customer session

mrmcduff-stripe opened the pull request:
r? @bg-stripe 
cc @ksun-stripe @joeydong-stripe 

This adds a CustomerSession class, along with EphemeralKeyProvider and EphemeralKeyUpdateListener (a callback interface).

Intentionally omitted functions/classes to be added in the future:
1 - an example activity that uses these (I have tested with one)
2 - a KeyManager class that updates the EphemeralKey on a schedule
3 - Checking to see if the EphemeralKey is "close enough" to expired. Right now, if there is only one millisecond left before the key expires, I call it valid.
4 - refreshing the customer object after a given period
5 - handling API errors

Note that testing multithreaded operations is not currently supported in our repo, so if a StripeApiProxy is added, all "networking" (it does not hit the network) is done on the main thread.

bg-stripe commented on stripe/src/main/java/com/stripe/android/StripeApiHandler.java:
```diff
@@ -63,11 +64,14 @@
     private static final String DNS_CACHE_TTL_PROPERTY_NAME = "networkaddress.cache.ttl";
     private static final SSLSocketFactory SSL_SOCKET_FACTORY = new StripeSSLSocketFactory();
 
+    static final String API_VERSION = "2015-10-12";
```
huh, so the android SDK wasn't explicitly setting an API version before? that's not great (by default it's the merchant's API version, which may not match the SDK's expectations.) can we add this to all requests the SDK makes?

joeydong-stripe replied:
maybe we can just start with the latest API version. we have this version in the iOS SDK because it was already there before we added ephemeral key support and weren't sure of the implications of changing it.

mrmcduff-stripe replied:
Heh. I definitely just copied it from iOS. Can update.

bg-stripe commented on stripe/src/main/java/com/stripe/android/CustomerSession.java:
```diff
@@ -0,0 +1,299 @@
+        return customer;
+    }
+
+    static boolean isTimeInPast(long milliSeconds, @Nullable Calendar nowCalendar) {
```
maybe better to call this something like `shouldRefreshKey`?

mrmcduff-stripe replied:
This method has no idea whether or not it's addressing a key, though -- isTimeWithinDeltaOfPresent(long time, long delta, Calendar debugCalendar) might make it more useful, since i'll need to check on the age of the customer in the future

bg-stripe replied:
oh, i meant you could change this to `shouldRefreshKey` and take a key as a parameter. (and move this stuff to `EphemeralKeyManager`)

bg-stripe commented on stripe/src/main/java/com/stripe/android/CustomerSession.java:
```diff
@@ -0,0 +1,299 @@
+     * @return a {@link Customer} if one can be found with this key, or {@code null} if one cannot.
+     */
+    @Nullable
+    static Customer retrieveCustomerWithKey(
```
This should be an instance method, so you can use the `CustomerSession`'s cached customer if it's fresh enough.

mrmcduff-stripe replied:
This method is basically just the call into the API -- I don't do checking on anything except whether or not a key is expired.

mrmcduff-stripe replied:
Also, this can't be a non-static method because it invokes the static method from the API class.

bg-stripe replied:
ah, never mind, i thought this was the public API for getting a customer – looks like it's `retrieveCurrentCustomer`, which looks fine

bg-stripe commented on stripe/src/main/java/com/stripe/android/StripeApiHandler.java:
```diff
@@ -63,11 +64,14 @@
     private static final String DNS_CACHE_TTL_PROPERTY_NAME = "networkaddress.cache.ttl";
     private static final SSLSocketFactory SSL_SOCKET_FACTORY = new StripeSSLSocketFactory();
 
+    static final String API_VERSION = "2017-06-05";
```
huh, so the android SDK wasn't explicitly setting an API version before? that's not great (by default it's the merchant's API version, which may not match the SDK's expectations.) can we add this to all requests the SDK makes?

mrmcduff-stripe replied:
We can, but that probably shouldn't be in this PR. Also, don't we want to hit the API that the user is expecting to hit, since that's what they'll be hitting with their own backend code?

bg-stripe replied:
So it's perfectly fine for the SDK and the user's backend to be on different API versions – the only exception to this rule is for the ephemeral key request, which is why we pass the SDK's API version to the user's backend. 

It's important for the SDK to be locked to an API version so that it doesn't receive unexpected responses that it can't parse, or pass unsupported parameters.

mrmcduff-stripe replied:
Ok. I'll see about putting in an api version dynamically as well - we should also probably start pushing up locales so that we get localized error messages (have a bug out on that one)

mrmcduff-stripe commented on stripe/src/main/java/com/stripe/android/PaymentConfiguration.java:
```diff
@@ -61,6 +62,15 @@ public PaymentConfiguration setShouldUseSourcesForCards(boolean shouldUseSources
         return this;
     }
 
+    void setEphemeralKeyProviderClassLoader(@NonNull ClassLoader classLoader) {
```
We would have to make CustomerSession a singleton to do that in order to keep the ClassLoader _somewhere_ if we don't just store its ClassLoader in our only singleton thus far. But if `CustomerSession` is a singleton, then I can shave off about half of this PR because we don't need to bother with Parcelable

mrmcduff-stripe replied:
This is actually part of my fud on calling `PaymentConfiguration` by that name -- it's really a `StripeConfiguration`, since it contains things that are useful beyond the `PaymentSession`

bg-stripe replied:
can we remove this now that `CustomerSession` is a singleton?

mrmcduff-stripe commented on stripe/src/main/java/com/stripe/android/CustomerSession.java:
```diff
@@ -0,0 +1,299 @@
+            @Nullable Calendar proxyNowCalendar) {
+        mKeyProvider = keyProvider;
+        mEphemeralKey = EphemeralKey.getEmptyKey();
+        mCustomer = Customer.getEmptyCustomer();
```
Not if we need this to be `Parcelable` -- though going down to the less-performant `Serializable` interface would make this easier. But if we're making this into a Singleton any way, then there's no need to ever pass it between Activities, which obviates the need to stuff it in an Intent, which means it doesn't matter if its parcelable or serializable.

mrmcduff-stripe commented:
Updates:
* `CustomerSession` is now a singleton, which means it is no longer `Parcelable`, and neither is `EphemeralKeyProvider`
* Added an `EphemeralKeyManager` class that handles retrieving and updating the key.
* Added the concept of a "time buffer" in the manager that we can set to whatever we want in the future (best location of this constant would be in `CustomerSession`)
* `CustomerSession` now assumes it has an up-to-date key (thanks to the key manager)

Intentionally not updated:
* the `Customer` field of `CustomerSession` does not go stale
* the `EphemeralKeyManager` does not yet have a constantly cycling alarm to update the key at certain intervals.

bg-stripe commented on stripe/src/main/java/com/stripe/android/EphemeralKeyManager.java:
```diff
@@ -0,0 +1,108 @@
+        return mEphemeralKey;
+    }
+
+    private void updateSessionKey(@NonNull String key) {
```
nit: the `Session` in these method names is kinda confusing

mrmcduff-stripe replied:
i wanted to emphasize whose key we're updating. It's actually not specifically talking to a CustomerSession, so maybe it should be a listener.

bg-stripe commented on stripe/src/main/java/com/stripe/android/EphemeralKeyManager.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    void retrieveEphemeralKey() {
+        if (mEphemeralKey != null &&
```
nit: you do a null check in `shouldRefreshKey` so could probably remove this

mrmcduff-stripe replied:
Ahh. Yeah, I think I used to try and access the time and only pass that to the expiry check function.

bg-stripe commented on stripe/src/main/java/com/stripe/android/model/Customer.java:
```diff
@@ -103,6 +103,11 @@ public JSONObject toJson() {
     }
 
     @NonNull
+    public static Customer getEmptyCustomer() {
```
looks like this is unused now

mrmcduff-stripe replied:
Oh, forgot to delete

bg-stripe commented on stripe/src/main/java/com/stripe/android/CustomerSession.java:
```diff
@@ -0,0 +1,276 @@
+
+    /**
+     * Get the currently cached {@link Customer}. Note that this may be an empty
+     * customer value (equivalent to {@link Customer#getEmptyCustomer()} if there is no
```
Note about `getEmptyCustomer` is stale now that customer is nullable
